### PR TITLE
Remove pipelining and sharding from LLM

### PR DESCRIPTION
**Pull request by rsuderman**
Pipelining and sharding is making it impossible to prototype and improve interfaces. Stripping them should let us rework pipelining in a more sustainable way and rework how the kvcache and attention layers work.

### CODE REVIEW

File: sharktank/tests/examples/paged_llm_v1_test.py [hunk 1]
```diff
@@ -19,17 +19,15 @@
...
+    "dtype",
     [
-        (torch.float16, 1),
-        (torch.float16, 2),
-        (torch.float32, 1),
-        (torch.float32, 2),
+        torch.float16,
+        torch.float16,
+        torch.float32,
+        torch.float32,
```
Comment (archana-ramalingam) [inline sharktank/tests/examples/paged_llm_v1_test.py:27]: Will need to remove the duplicates

File: sharktank/tests/models/llama/test_llama.py [hunk 2]
```diff
@@ -65,33 +63,21 @@ def testUnsharded(self):
...
-                iree_hip_target=self.iree_hip_target,
-                iree_hal_target_device=self.iree_hal_target_device,
-            )
-            tester.run_and_compare_iree_vs_eager()
-        except IreeCompileException as e:
-            if tensor_parallelism_size == 2 or pipeline_parallelism_size == 2:
-                pytest.xfail(reason="sharding compilation is broken")
-            else:
-                raise e
+        theta, config = generate(12345)
```
Comment (archana-ramalingam) [inline sharktank/tests/models/llama/test_llama.py:73]: Remove duplicate

File: sharktank/sharktank/models/llm/llm.py [hunk 3]
```diff
@@ -215,55 +173,21 @@ def prefill(
...
-        embedding_batch_masks: list[tuple[InferenceTensor, InferenceTensor]] | list[
-            InferenceTensor
-        ] = []
-        for pipeline, start_position in enumerate(start_positions):
-            mask = self.attention_embedding[pipeline].compute_batch_mask(
-                start_position, batch_seq_len=1
-            )
-            embedding_batch_masks.append(mask)
-            # TODO: How to name and trace this properly
-            self.trace_tensor("llama.embedding_batch_mask", mask)
```
Comment (archana-ramalingam) [inline sharktank/sharktank/models/llm/llm.py:266]: Might want to add this back to trace `embedding_batch_masks`

File: sharktank/sharktank/models/llm/llm.py [hunk 4]
```diff
@@ -122,42 +114,17 @@ def __init__(self, theta: Theta, config: LlamaModelConfig):
...
-        seq_block_ids: list[Union[torch.Tensor, ReplicatedTensor]],
-        cache_state: list[Union[torch.Tensor, SplitPrimitiveTensor]],
+        seq_block_ids: torch.Tensor,
+        cache_state: torch.Tensor,
     ):
-        self._assert_device(tokens)
-        if not all(mask is None for mask in attention_mask):
-            self._assert_device(*attention_mask, dtype=self.activation_dtype)
-        self._assert_device(*seq_block_ids)
-        self._assert_device(*cache_state, dtype=self.activation_dtype)
```
Comment (archana-ramalingam) [inline sharktank/sharktank/models/llm/llm.py:160]: Do we not require these asserts for unsharded models?

**Review by archana-ramalingam — CHANGES_REQUESTED** [review]
Other TP/PP code that needs to be removed:
- evaluate/perplexity_iree.py
- evaluate/perplexity_torch.py
- utils/export_artifacts.py
- layers/rotary_embedding.py
- tests/layers/mixture_of_experts_block_test.py
- tests/models/llama/benchmark_amdgpu_test.py

**Comment by sogartar** [comment]
@rsuderman Could you go into a bit more detail what issues did you encounter regarding sharding prevented you from implementing the new feature?

File: sharktank/tests/models/llama/test_llama.py (hunk 2 above)
Comment (rsuderman) [reply sharktank/tests/models/llama/test_llama.py]: Fixed.

File: sharktank/tests/evaluate/perplexity_iree_test.py [hunk 5]
```diff
@@ -132,58 +129,6 @@ def test_llama3_8B_f8(self):
         )
         self.run_and_check_perplexity()
 
-    @is_sharded
```
[the remark this answers is not in the document]
Comment (rsuderman) [reply sharktank/tests/evaluate/perplexity_iree_test.py:135]: See rerequest comment.

**Comment by rsuderman** [comment]
> Other TP/PP code that needs to be removed:
> 
> * evaluate/perplexity_iree.py
> * evaluate/perplexity_torch.py
> * utils/export_artifacts.py
> * layers/rotary_embedding.py
> * tests/layers/mixture_of_experts_block_test.py
> * tests/models/llama/benchmark_amdgpu_test.py

I tried to avoid changing sections where sharding was either inconsequential (rotary_embedding) or areas that are going through significant rewrites anyway. E.g. I have a separate PR with new framework for numerical evaluation that I want to hookup as a replacement for a perplexity work. I am hoping we can drastically simplify the perplexity evaluations then delete the current implementation entirely.

File: sharktank/tests/models/llama/test_llama.py [hunk 6]
```diff
@@ -65,33 +63,20 @@ def testUnsharded(self):
...
         strict=True,
     )
-    def testUnshardedToyIreeVsEager(
-        self, tensor_parallelism_size: int, pipeline_parallelism_size: int, use_hf: bool
-    ):
+    def testUnshardedToyIreeVsEager(self):
         theta, config = generate(12345)
-        config.tensor_parallelism_size = tensor_parallelism_size
-        config.pipeline_parallelism_size = pipeline_parallelism_size
-        config.use_hf = use_hf
```
Comment (archana-ramalingam) [inline sharktank/tests/models/llama/test_llama.py:80]: Is there a reason we don't want to parametrize this test for `use_hf` too?

**Review by archana-ramalingam — APPROVED** [review]

File: sharktank/tests/models/llama/test_llama.py (hunk 6 above)
Comment (rsuderman) [reply sharktank/tests/models/llama/test_llama.py:80]: Overall we shouldn't be performing numerical validation this (squared error difference on logits). It will validate that they do not deviate from compilation error but thats a pretty minimal correctness benefit. The longterm plan is to replace this with cross entropy validation for IREE as well so just consider this a continued tear out of old testing.
